Index: ffmpeg-8.0/libavcodec/h2645_sei.c
===================================================================
--- ffmpeg-8.0.orig/libavcodec/h2645_sei.c
+++ ffmpeg-8.0/libavcodec/h2645_sei.c
@@ -915,5 +915,7 @@ void ff_h2645_sei_reset(H2645SEI *s)
     s->content_light.present = 0;
 
     av_refstruct_unref(&s->film_grain_characteristics);
+#if CONFIG_HEVC_SEI
     ff_aom_uninit_film_grain_params(&s->aom_film_grain);
+#endif
 }
