diff --git a/PATCHED b/PATCHED
new file mode 100644
index 0000000..e69de29
diff --git a/codec/api/svc/codec_api.h b/codec/api/svc/codec_api.h
index a1326c8..7d57e0f 100644
--- a/codec/api/svc/codec_api.h
+++ b/codec/api/svc/codec_api.h
@@ -486,7 +486,7 @@ int (*Uninitialize) (ISVCEncoder*);
 int (*EncodeFrame) (ISVCEncoder*, const SSourcePicture* kpSrcPic, SFrameBSInfo* pBsInfo);
 int (*EncodeParameterSets) (ISVCEncoder*, SFrameBSInfo* pBsInfo);
 
-int (*ForceIntraFrame) (ISVCEncoder*, bool bIDR);
+int (*ForceIntraFrame) (ISVCEncoder*, bool bIDR, int iLayerId);
 
 int (*SetOption) (ISVCEncoder*, ENCODER_OPTION eOptionId, void* pOption);
 int (*GetOption) (ISVCEncoder*, ENCODER_OPTION eOptionId, void* pOption);
