diff --git a/codec/api/wels/codec_api.h b/codec/api/wels/codec_api.h
index a1326c8..7d57e0f 100644
--- a/codec/api/wels/codec_api.h
+++ b/codec/api/wels/codec_api.h
@@ -486,7 +486,7 @@ int (*Uninitialize) (ISVCEncoder*);
 int (*EncodeFrame) (ISVCEncoder*, const SSourcePicture* kpSrcPic, SFrameBSInfo* pBsInfo);
 int (*EncodeParameterSets) (ISVCEncoder*, SFrameBSInfo* pBsInfo);
 
-int (*ForceIntraFrame) (ISVCEncoder*, bool bIDR);
+int (*ForceIntraFrame) (ISVCEncoder*, bool bIDR, int iLayerId);
 
 int (*SetOption) (ISVCEncoder*, ENCODER_OPTION eOptionId, void* pOption);
 int (*GetOption) (ISVCEncoder*, ENCODER_OPTION eOptionId, void* pOption);
diff --git a/codec/common/src/WelsThreadLib.cpp b/codec/common/src/WelsThreadLib.cpp
index 4477e34..37f0adf 100644
--- a/codec/common/src/WelsThreadLib.cpp
+++ b/codec/common/src/WelsThreadLib.cpp
@@ -48,9 +48,6 @@
 #include <sys/types.h>
 #include <sys/param.h>
 #include <unistd.h>
-#ifndef __Fuchsia__
-#include <sys/sysctl.h>
-#endif
 #ifdef __APPLE__
 #define HW_NCPU_NAME "hw.logicalcpu"
 #else
diff --git a/codec/decoder/core/src/wels_decoder_thread.cpp b/codec/decoder/core/src/wels_decoder_thread.cpp
index d05aa45..7eafe90 100644
--- a/codec/decoder/core/src/wels_decoder_thread.cpp
+++ b/codec/decoder/core/src/wels_decoder_thread.cpp
@@ -48,9 +48,6 @@
 #include <sys/types.h>
 #include <sys/param.h>
 #include <unistd.h>
-#ifndef __Fuchsia__
-#include <sys/sysctl.h>
-#endif
 #ifdef __APPLE__
 #define HW_NCPU_NAME "hw.logicalcpu"
 #else
@@ -249,8 +246,10 @@ int SemWait (SWelsDecSemphore* s, int32_t timeout) {
 #if defined(__APPLE__)
         rc = pthread_cond_timedwait (& (s->e), & (s->m), &ts);
 #else
+#if !defined(__EMSCRIPTEN__) || defined(__EMSCRIPTEN_PTHREADS__)
         rc = sem_timedwait (s->e, &ts);
         if (rc != 0) rc = errno;
+#endif
 #endif
         if (rc != EINTR) {
           // if timed out we return to the caller
